# Finding Profitable App Genre - Google Play & App Store Market

My goal in this project is to find the genre(ex: Games, Books, News etc) of app's that will be profitable on both Google Play and the iOS App Store markets.This will help freelance app developers to make better data driven decision on the kind of apps they would like to develop to be profitable.

We will analyse data about Google Play and App Store apps and try to find:

* Most common apps by genre on both the markets
* Most popular apps by genre on both the markets

### Summary of Results

After analyzing the data, we found that taking a recent popular book and turning it into an app will be profitable on both the markets. We also recommend adding several features to the app. For more details, please refer to the full analysis below.


## Data Set

Both Google Play and App Store have more than 2 million apps each. Gathering data for these 4+ million apps will be resource intense and hence we decided to analyse sample of data that we located on Kaggle.

* This [data set](https://www.kaggle.com/lava18/google-play-store-apps) contains data about 10,000+ Android apps on the Google Play.

* This [data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) contains data about 7,000+ iOS apps on the App Store.

## Exploring the Data set

We will start by opening the two data set files that we are going to analyse.

In [1]:
#Opening the Google Play data set
open_android = open("googleplaystore.csv")
from csv import reader
android_reader = reader(open_android)
android = list(android_reader)  

#Opening the iOS App Store data set
open_ios = open("AppleStore.csv")
ios_reader = reader(open_ios)
ios = list(ios_reader)



 To make our analysis easier, let's write a small function named app_info that will return the app information we want for any data set. Apart from it, this function will also return the total number of app's in that particular data set if needed. 

In [2]:
#Function to return app information
def app_info(dataset, start, finish, length= False):
    for each_app in dataset[start:finish+1]:
        print(each_app)
        print("\n")
        
    if length == True:
        print("Total number of apps: ", len(dataset[1:]))
        

We will use the above function to print few apps from both the data set to see how the information looks like.

In [3]:
##Using app_info function to print Android apps
app_info(android, 0, 3, True)


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Total number of apps:  10841


As we above, there are total of 10,841 Android apps. The app contains information like name, size, number of reviews, installs, average rating to name a few.

Now lets see how may iOS apps we have for analysis and see how it looks like.

In [ ]:
#Using app_info function to print iOS apps
app_info(ios, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Total number of apps:  7197


We have 7197 iOS apps in our App Store data set and it contains information's like name, size, price, total number of ratings, average rating, genre etc.

One thing to note in App Store data set is that, it does not contain number of install information like we had for Google Play data set. 

## Data Cleaning

 The most underrated process in a data analysis task is cleaning of data. Before we start analyzing our data, we need to be sure that the data is free of errors, misinformation's and duplicates. Else our analysis would be inaccurate and can lead to false conclusions. So it is very important that we take time to clean and make our data set ready for further analysis. 

### Remove Inaccurate Data

First step in the data cleaning process is to find out if there are app's in the data set that has any missing information.

In [ ]:
##Find inaccurate app in Google Play
for each in android[1:]:
    if len(each) != len(android[0]):
        print(each)
        print("length of this row: ", len(each))
        print("Index of inaccurate app is: ", android.index(each))
        print("\n")
        print(android[0])
        print("length of the header: ", len(android[0]))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
length of this row:  12
Index of inaccurate app is:  10473


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
length of the header:  13


As we see, information of the app named "Life Made WI-Fi Touchscreen Photo Frame" on the Google Play data set is inaccurate as it is missing one of the information. The app has only 12 columns against the 13 we have for the header of Google Play data set. So we better of remove this app as we dont know which information is missing.

In [ ]:
## Removing "Life Made WI-Fi Touchscreen Photo Frame" app
print("Number of apps before deleting: ", len(android[1:]))
del android[10473] 
print("Number of apps after deleting: ",len(android[1:]))

Number of apps before deleting:  10841
Number of apps after deleting:  10840


In [ ]:
##Find inaccurate app information in App Store
for each in ios[1:]:
    if len(each) != len(ios[0]):
        print(each)
        print("length of this row: ", len(each))
        print("Index of inaccurate app is: ", ios.index(each))
        print("\n")
        print(ios[0])
        print("length of the header: ", len(ios[0]))

There is no app(s) on the App Store that is missing information.

### Deleting Duplicate App(s)

Next step in data cleaning is to locate apps that occur more than ones and find a way to remove those duplicates. For example the app "Google Ads" occur thrice in Google Play data set as we see below.

In [ ]:
for each_app in android[1:]:
    name = each_app[0]
    if name == "Google Ads":
        print(each_app)

['Google Ads', 'BUSINESS', '4.3', '29313', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 30, 2018', '1.12.0', '4.0.3 and up']
['Google Ads', 'BUSINESS', '4.3', '29313', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 30, 2018', '1.12.0', '4.0.3 and up']
['Google Ads', 'BUSINESS', '4.3', '29331', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 30, 2018', '1.12.0', '4.0.3 and up']


Now lets find how many duplicate entries are there in each of the data sets.

In [ ]:
## Finding number of duplicate apps in the Google Play
duplicate_app_google = []
unique_app_google = []
for each_row in android[1:]:
    name = each_row[0]
    if name in unique_app_google:
        duplicate_app_google.append(name)
    else:
        unique_app_google.append(name)
        
print("Number of duplicate apps in Google Play: ", len(duplicate_app_google))
print("\n")
print("Number of unique apps in Google Play: ", len(unique_app_google))

## Finding number of duplicate apps in the App Store
duplicate_app_apple = []
unique_app_apple = []
for each_row in ios[1:]:
    name = each_row[1]
    if name in unique_app_apple:
        duplicate_app_apple.append(name)
    else:
        unique_app_apple.append(name)

print("\n")        
print("Number of duplicate apps in App Store: ", len(duplicate_app_apple))
print("\n")
print("Number of unique apps in App Store: ", len(unique_app_apple))


        

We can see that there are 1181 duplicate apps in the Google Play data set. App Store is much better and contains only 2 duplicate entries.

We need to keep only 1 entry per app and delete all the duplicate ones. If you see the "Google Ads" we printed above, there are 2 duplicates and the difference between the apps happen at 4th column which is the total number of reviews. So we can keep the app that has the highest number of reviews and delete the remaining ones. Higher the number of reviews the latest the information of the app should be.

We will also follow the same procedure for App Store data set to keep the unique apps.

Lets start by creating a dictionary where keys will be the unique app name and the values will be the highest review count for that app.

In [ ]:
#Finding number of unique apps in Google Play
google_unique_app = {}
for each_row in android[1:]:
    name = each_row[0]
    reviews = float(each_row[3])
    if name in google_unique_app and reviews > google_unique_app[name]:
        google_unique_app[name] = reviews
    elif name not in google_unique_app:
        google_unique_app[name] = reviews
        
print("Number of unique Google apps extracted: ", len(google_unique_app))

As we can see, the number of unique apps in the dictionary is 9659 which is same as what we found in the previous code cell.

Now let us use this dictionary to create a list, with all the rows of the unique apps having the highest number of reviews.

In [ ]:
unique_google = []
already_added = []
for each_row in android[1:]:
    name = each_row[0]
    review = float(each_row[3])
    if google_unique_app[name] == review and name not in already_added:
        unique_google.append(each_row)
        already_added.append(name)

print("Number of Google Play apps :", len(unique_google))
        

Now lets do the same code for App Store data set and extract the unique app having the highest review. The number of reviews is found at column 6 in the App Store data set and hence we will modify our code accordingly.

In [ ]:
apple_unique_app = {}
for each_row in ios[1:]:
    name = each_row[1]
    reviews = float(each_row[5])
    if name in apple_unique_app and reviews > apple_unique_app[name]:
        apple_unique_app[name] = reviews
    elif name not in apple_unique_app:
        apple_unique_app[name] = reviews
        
print("Number of unique iOS apps extracted: ", len(apple_unique_app))

unique_apple = []
already_added = []
for each_row in ios[1:]:
    name = each_row[1]
    review = float(each_row[5])
    if apple_unique_app[name] == review and name not in already_added:
        unique_apple.append(each_row)
        already_added.append(name)

print("Number of App Store apps :", len(unique_apple))
        

### Removing Non-English Apps

There are several non-english apps in both the data set. For example:

['Cъновник BG', 'BOOKS_AND_REFERENCE', 'NaN', '13', '4.1M', '1,000+', 'Free', '0', 'Everyone', 'Books & Reference', 'January 21, 2017', '250', '4.0 and up']

['뽕티비 - 개인방송, 인터넷방송, BJ방송', 'VIDEO_PLAYERS', 'NaN', '414', '59M', '100,000+', 'Free', '0', 'Mature 17+', 'Video Players & Editors', 'July 18, 2018', '4.0.7', '4.0.3 and up']

['BL 女性向け恋愛ゲーム◆俺プリクロス', 'FAMILY', '4.2', '3379', '62M', '100,000+', 'Free', '0', 'Mature 17+', 'Simulation', 'March 23, 2017', '1.6.3', '2.3.3 and up']

This project mainly caters to English speaking audience and hence it makes no sense to analyse app's that are not in English. So as part of data cleaning, we will remove all the non-english apps.

Each alphabet we type in computer corresponds to a number called "ASCII" value. The texts we normally use in English(letters, numbers, punctutaions and other symbols) sits in the ASCII range of 0 - 127.

The logic is to go through each row on both the data sets and extract the name of the app and check if the ASCII value of each of their characters are > 127. if it's greater than 127 we will remove those apps.

However certain English apps still have some characters in their name whose ASCII value are greater than 127. For example:

['**FlirtChat - ♥Free Dating/Flirting App♥**', 'DATING', '4.3', '2433', '13M', '500,000+', 'Free', '0', 'Mature 17+', 'Dating', 'July 26, 2018', '12.0.4', '4.1 and up']

if you see the app above,  we are sure that the app is English but the ASCII value of  two ♥ characters we have in the name, is 9829. So going by our rule, our code will also remove apps like these, since one of the characters in the name field has ASCII value greateer than 127.

So we will modify our code in such a way that, our program will remove the apps only if the app name have more than 3 ASCII characters whose values are greater than 127. By this way we can keep most of the english apps that have special characters in their names like we saw above. 


In [ ]:
##Removing non-English app from the Google Play
google_english = []
for each_app in unique_google:
    name = each_app[0]
    count = 0
    for each in name:
        if ord(each) > 127:
            count += 1
    if count <= 3:
        google_english.append(each_app)
        
print("Number of English apps in the Goole Play: ", len(google_english))

As we see, against the original 9659 apps we are left with 9614 apps after removing the non-english apps. Now lets do the same for apps in the App Store

In [ ]:
##Removing non-English app from the App Store
ios_english = []
for each_app in unique_apple:
    name = each_app[1]
    count = 0
    for each in name:
        if ord(each) > 127:
            count += 1
    if count <= 3:
        ios_english.append(each_app)
        
print("Number of English apps in the App Store: ", len(ios_english))

So we have removed quite a few apps from the App Store and finally left with 6181 english apps.

Our code may still have few non english apps and there is also a possibility that we might have left few english apps, but that should not have any significant bearing in the outcome of our analysis.

### Removing paid Apps

Almost all the freelancers that we know of, wants to build only free apps since that will bring in more number of users against a paid one for which users expectations will be higher and also require significant maintenance costs. Most of the freelancers are students and hence they prefer developing a free app.

So success of any free app depends up on the number of users downloads it and the revenue comes form the in-app ads. So more the number of users download and use the app, more is the revenue.

So for the reasons said above, we will remove the paid apps from both the data sets.


In [ ]:
## Removing paid apps from the google data set
android_final = []
for each_app in google_english:
    if each_app[7] == "0":
        android_final.append(each_app)

print("Number of Android apps after cleaning: ", len(android_final))

## Removing paid apps from the apple data set
ios_final = []
for each_app in ios_english:
    if each_app[4] == "0.0":
        ios_final.append(each_app)
print("Number of iOS apps after cleaning: ", len(ios_final))    

After removing the paid apps, we are left with 8864 Android apps and 3220 iOS apps.

With this, the data cleaning process is complete.

## Data Analysis

As said in the introduction, our goal is to find the type of apps that will be profitable on both the markets. Since we are dealing only with free apps, the revenue of any free apps depends on the number of users using the app.

### Most Common App by Genre

So lets start our data analysis by.finding what genre of apps are very common on two markets. The genre information for Android is on index 9 and for iOS it is on index 11.

We will be creating two functions ; one to create a frequency table for any rows we want and the next function to sort and print the frequenct table in descending order so that it wil be readable.

In [ ]:
## Function to create frequency table
def freq_table(dataset, index):
    frequency_table = {}
    for each_app in dataset:
        column = each_app[index]
        if column in frequency_table:
            frequency_table[column] += 1
        else:
            frequency_table[column] = 1
            
    for each_item in frequency_table:
        frequency_table[each_item] /= len(dataset)
        frequency_table[each_item] *= 100
        
    return frequency_table

## Function to sort the above frequency table in descending order
def sort(dataset, index):
    frequency_table = freq_table(dataset, index)
    list_freq = []
    for each in frequency_table:
        freq_temp = [frequency_table[each], each]
        list_freq.append(freq_temp)
        
    sort_freq = sorted(list_freq, reverse = True)
    for each in sort_freq:
        print(each[1], " : ", each[0])

In [ ]:
sort(ios_final, 11)

As we see, among the Free English apps more than half(58 %) of them belongs to *Games* category, followed by *Entertainment* at a distant second (~ 8 %). *Photo & Video* takes 3rd place with ~ 9%. 

So we can see that almost 70% of the apps belongs to fun category(Games, Entertainment, Photos).

Now lets take a look at Android market.

In [ ]:
sort(android_final, 9)

Unlike App Store which are dominated by Fun apps, The Google Play store has balanced landscape with both fun(entertainment, lifestyle) and apps designed for productivity purposes(tools, education) in the top.

However, we also need to understand that on both Google Play and App Store markets, genres with most proportion may or may not have highest number of users(downloads) as well. The demand might not be same as supply.

### App with most users

Now lets figure out the genre of apps with most users.

If you see the App Store data set there is no column that gives the information about the number of users/downloads. So what we will do instead is use column number 5, which is the total number of reviews which kind of gives us an idea about its popularity.


In [ ]:
## Finding Apps of Genre with most users on App Store
most_apps = {}
temp_add = {}
for each_app in ios_final:
    rating_count = float(each_app[5])
    genre = each_app[11]
    if genre in most_apps:
        most_apps[genre] *= temp_add[genre]
        most_apps[genre] += rating_count
        temp_add[genre] += 1
        most_apps[genre] /= temp_add[genre]
    
    else:
        most_apps[genre] = rating_count
        temp_add[genre] = 1

as_list = []
for each in most_apps:
    temp_list = [most_apps[each], each]
    as_list.append(temp_list)
    
for each_app in sorted(as_list, reverse = True):
    print(each_app[1], " : ", each_app[0])
    

*Navigation* apps has the highest number of reviews on average. However after analysing further in detail, we could see that the average was due to *Waze* and *Google Maps* which has more than 95% of total reviews.

In [ ]:
for each in ios_final:
    if each[11] == "Navigation":
        print(each[1], " :", each[5])
        print("\n")

We notice the same trend in *Social Networking* and *Music* genre as well. Social Networking is dominated by Facebook and Pinterest and Music category is dominated by Pandora and Spotify. It is impossible for a freelancer to compete against these giants in this category and succeed, because all these are very successful products with millions of users in real time. Same thing goes for *Food* and *Finance* category which is dominated by giants in those field.

Reference category has 74942 reviews on average. On looking further, Bible and Dictionary tops the list with most reviews.

In [ ]:
for each in ios_final:
    if each[11] == "Reference":
        print(each[1], " :", each[5])
        print("\n")

This catgory seems interesting. What we can do is take one or more popular book and turn into an app. We can add many different features like quiz, quotes, links to any news aboust the book on the web. We also can include audio version of the book.

Another interesting thing we can do is to have an in built Dictionary inside the app itself so the user dont have to exit the app to look for meaning.

Now lets find in Goole Play the genre of apps with most users.

In [ ]:
## Finding Apps of Genre with most users on Google Play
most_apps = {}
temp_add = {}
for each_app in android_final:
    rating_count = each_app[5]
    rating_count = rating_count.replace("+", "")
    rating_count = rating_count.replace(",", "")
    rating_count = float(rating_count)
    genre = each_app[9]
    if genre in most_apps:
        most_apps[genre] *= temp_add[genre]
        most_apps[genre] += rating_count
        temp_add[genre] += 1
        most_apps[genre] /= temp_add[genre]
    
    else:
        most_apps[genre] = rating_count
        temp_add[genre] = 1

as_list = []
for each in most_apps:
    temp_list = [most_apps[each], each]
    as_list.append(temp_list)
    
for each_app in sorted(as_list, reverse = True):
    print(each_app[1], " : ", each_app[0])

*Communication* tops the list with average install of 38 million. However same like App Store market, most number of installs belong to  *WhatsAPP*, *Skype*, *Messenger* and few more apps which have more than 1 billion installs.

In [ ]:
for each in android_final:
    if each[9] == "Communication":
        rating = each[5]
        rating = rating.replace("+", "")
        rating = rating.replace(",","")
        rating = float(rating)
        if rating > 100000000:
            print(each[0], " : ", each[5])

So *Communication* Apps seems to be more popular than they really are. If you exclude the Communications apps with more than 10 million installs, the average installs reduce to 0.7 million.

In [ ]:
#average rating by excluding > 10 million apps
ratings = []
for each in android_final:
    if each[9] == "Communication":
        rating = each[5]
        rating = rating.replace("+", "")
        rating = rating.replace(",","")
        rating = float(rating)
        if rating < 10000000:
            ratings.append(rating)

print("Average installs of communication apps ater excluding apps greater than 10 million installs: ",sum(ratings) / len(ratings))

The Genre after *Communication* is *Adventure;Action & Adventure* which only have 3 apps which dont give much information for our analysis and hence we skip that.

In [ ]:
for each in android_final:
    if each[9] == "Adventure;Action & Adventure":
        print(each[0], " : ", each[5])

The other top Genres in the list, also follows similar pattern like Communications apps where only the few giants in the field have almost all the installs. For example *Youtube*  and *Google Play* have more than *1 billion* in the *Video Players & Editors Genre*.

In [ ]:
ratings = []
for each in android_final:
    if each[9] == "Video Players & Editors":
        rating = each[5]
        rating = rating.replace("+", "")
        rating = rating.replace(",","")
        rating = float(rating)
        if rating > 100000000:
            print(each[0], " : ", each[5])

*Books & Reference* genre have nearly 9 million installs. We would like to explore this in detail as we found this genre has some potential in App Store, and hence would like to see how it fairs in the Google Play market.`


In [ ]:
for each in android_final:
    if each[9] == "Books & Reference":
        print(each[0], " : ", each[5])

Books & Reference section seems to have varierty of apps like Ebook, Dictionary, programming language etc.

We also see successful apps built around Bible and Quran which suggests that taking a recent popular book and turning into an app will be profitable.

Instead of having a raw version of the book, we should add different features to make it more interesting. For example: quizzes, quote of the day, news and interviews regarding the book, audio version and an inbuild dictionary to name a few.

## Conclusion

In this project, we analysed data about Google Play and App Store inorder to recommend the genre of apps which will be successful on both markets.

Baes on our analysis we recommend taking a recent popular book and turning into an app will be profitable. Inorder to make the app more appealing to users, we should add different features like audio version of the book, daily quizzes from the book, link to news item or interviews about the book, in-built dictionary and a forum inorder to encourage discussion about the book.